### PR TITLE
fix: only swallow connection errors #1304

### PR DESCRIPTION
**Pull request by tnolet**
Please reference the issue #1304

### CODE REVIEW

**Comment by tnolet** [comment]
This seems to only fail on Node 12

**Comment by manast** [comment]
ok, so we are still swallowing the error (i,.e. not throwing it up), it is just that we not write the error on the console.

**Comment by tnolet** [comment]
@manast I don't really know what you mean then. 

> ok, so we are still swallowing the error (i,.e. not throwing it up), it is just that we not write the error on the console.

So we do want to catch the "connection closed" error, but just rethrow it? i.e.
```
if (err.message === 'Connection is closed.') {
  throw(err)
}
```

Or do we want to ignore the "connection closed" error and rethrow all others?

```
if (err.message !== 'Connection is closed.') {
  throw(err)
}
```

in both cases, we log nothing.

**Comment by tnolet** [comment]
@manast would love to move ahead with this. We are still at risk blowing up our logging infrastructure if we have an hick up

**Comment by tnolet** [comment]
@manast It has been one month since I last pinged this PR. Can you let me know if there is any chance we this will be picked up by on of the maintainers?

**Comment by manast** [comment]
I am not sure what to say, this PR does nothing particularly useful, just changes when to console.log the error, it is not addressing #1304 as far as I can see.

**Comment by tnolet** [comment]
@manast Ok, clear. But it seems we are going in circles here a bit. What do we need to change to address #1304 effectively. As mentioned in the original issue:

###
@manast what is your suggestion for moving this ticket forward? I have the feeling we are going in circles a bit. I would be happy to remove/close my PR in favour of a better solution.

The reason I'm pushing this is that it is still a risk for many folks using logging solutions with hard limits.

As mentioned, I'm ready to help or have my team also look at it.
###

**Comment by manast** [comment]
I think that we need to add a delay, exponential best but fixed will also help, where in case of this "unexpected errors" we wait a couple of seconds. The connection is closed should not be handled as unexpected though since this is the error you get when closing the queue and you do not want to delay the closing more than necessary.

File: lib/queue.js [hunk 1]
```diff
@@ -1091,8 +1091,8 @@ Queue.prototype.getNextJob = function() {
           }
         },
         err => {
-          // Swallow error
-          if (err.message !== 'Connection is closed.') {
+          // Swallow error only for connection closed. See GH issue #1304
+          if (err.message === 'Connection is closed.') {
             console.error('BRPOPLPUSH', err);
           }
```
Comment (dobesv) [inline lib/queue.js:1097]:
```suggestion
          // Ignore 'Connection is closed.' error if closing the queue
          if (!(err.message === 'Connection is closed.' && this.closing)) {
            console.error('BRPOPLPUSH', err);
            throw err;
          }
```

Comment (tnolet) [reply lib/queue.js:1097]: @dobesv this is interesting. How about the delay mentioned by @manast? Do we need that, or not? It seems to be part of your suggestion here https://github.com/OptimalBits/bull/issues/1746

Comment (dobesv) [reply lib/queue.js:1097]: Yeah, after I proposed this I also thought of a way to sort of throttle attempts if we got an error.  It's hard to say exactly what the best policy is here, we're a bit in the dark.  But for a resilient system we could try to assume that redis is temporarily down or a graceful shutdown was initiated but we haven't been told yet (e.g. closing was not set *yet* but might be very soon).

It seems like maybe some kind of `emit('error')` or something like that to allow the library user to detect this condition might also be useful.

**Comment by tnolet** [comment]
This was fixed by https://github.com/OptimalBits/bull/pull/1974
Thanks to @manast for clearing this up!

**Pull request closed without merging**